### PR TITLE
Bookstore

### PR DESCRIPTION
**Pull request by Emmy-github-webdev**
# Bookstore

In this project, I:
- Ensure there is no lint error
- Follow Gitflow
- Refactor add book and remove a book
- Implement middleware to make an async requests to the API

### CODE REVIEW

File: src/redux/books/Books.js [hunk 1]
```diff
@@ -13,13 +16,34 @@ export const removeBook = (payload) => ({
...
   switch (action.type) {
     case ADD_BOOK:
+      newBook(action.payload);
       return [...state, action.payload];
 
     case REMOVE_BOOK:
+      deleteBook(action.payload);
       return state.filter((book) => book.id !== action.payload);
+    case GET_BOOK:
+      return [...state, action.payload];
```
Comment (D4nielJ) [inline src/redux/books/Books.js:46]: - [ ] I think one important rule about reducers and their cases it's that it must remain as a pure function, to achieve this we avoid generating side effects from it and we get the same result each time with the same input. Think "What happens if fetching the API is not available?"

That's why in this case it is recommended to use thunks to handle all async-related logic to our redux. Let's move these two functions (`newbook` and `deleteBook`) to a specific thunk and keep our reducer pure. [Read more about pure functions here.](https://www.sitepoint.com/functional-programming-pure-functions/#:~:text=A%20pure%20function%20is%20a,it%20does%20not%20change%20x%20.)

File: src/redux/books/Books.js [hunk 2]
```diff
@@ -13,13 +16,34 @@ export const removeBook = (payload) => ({
...
+  getBooks().then((books) => {
+    Object.keys(books).forEach((ID) => {
+      const booksId = {};
+      booksId.id = ID;
+      booksId.title = books[ID][0].title;
+      booksId.category = books[ID][0].category;
+      dispatch(loadBook(booksId));
+    });
+  });
+};
```
Comment (D4nielJ) [inline src/redux/books/Books.js:34]: A good example of creating an asynchronous thunk!

**Review by D4nielJ — CHANGES_REQUESTED** [review]
## Hello @Emmy-github-webdev, 

You're doing great! 👏 
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights :fire: :rocket: 
- The API stores correctly our data!
- Async thunk to `loadBooks` is well constructed.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
- Adding an option that allows the user to pick the category of the book. This is reflected on the design template and the API is adjusted to receive this information. 

## Happy coding and stay awesome! 🚀 

Feel free to leave any questions or comments in the PR thread if something is not 100% clear or you can always find me on slack!

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: src/redux/books/Books.js (hunk 2 above)
Comment (Emmy-github-webdev) [reply src/redux/books/Books.js:34]: Thanks, @D4nielJ 

Thunk is already used in store

image

Comment (D4nielJ) [reply src/redux/books/Books.js:34]: Hello Emmy, to clarify: Thunk is a middleware that allows the store to dispatch Asynchronous Functions named thunks. I'm not asking you to install the plugin again, it is correctly installed. 

I'm asking you to remove both async functions inside of the reducer to Thunk Functions, just as the one highlighted. [I hope this is more clear and if you need to read more about thunks you can do it here.](https://redux.js.org/usage/writing-logic-thunks)

Comment (Emmy-github-webdev) [reply src/redux/books/Books.js:34]: Asynchronous is implemented in API 

image

**Review by arslanbisharat — APPROVED** [review]
# Approved 👏 

Hi there, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

To Highlight: 
- The Pull Request has a proper title and description. ✔️ 
- You have added a descriptive Readme file. Good Job. ✔️ 
- There is no linter error is present. ✔️ 
- Project has met all the technical aspects you needed for this project. Keep rocking! ✔️ 

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of code reviews per project which you can see from your dashboard. If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)_

**Comment by Emmy-github-webdev** [comment]
Thanks, @arslanbisharat

**Pull request merged**